### PR TITLE
fix: don't mutate Run class when making disabled run

### PR DESCRIPTION
**Pull request by timoffex**
Moves the `WANDB_MODE=disabled` logic out of `wandb_init.py` and into its own file, mostly copying it over, but changing it to subclass `wandb.Run` instead of monkeypatching an instance of it. Subclassing works exactly the same, but avoids this horrible code in the original implementation:

```python
for attr in ("url", "project_url", "sweep_url"):
    setattr(type(drun), attr, property(lambda _: None))
```

`type(drun)` is `wandb.Run`, so this snippet changes how `wandb.Run` works after `wandb.init(mode="disabled")`. This shows up in test flakes like [this one](https://app.circleci.com/pipelines/github/wandb/wandb/64307/workflows/cdb8c0a4-ec07-4b38-9d14-e82c652830a0/jobs/1672741/tests).

The other horrible code is ported over as-is.

I spent a while trying to do this properly but gave up after too many edge cases. The least intrusive approach (not involving changes to `wandb_run.py`) is to semi-dynamically construct `NoopRun` (using reflection + hand-written overrides) and then write tests that inspect `wandb.Run` and ensure `NoopRun` has the desired interface: same public methods, dunders, properties, and attributes (like `run.summary`). But while writing it, I realized I'm not sure what guarantees we want to provide. The entire thing is sort of patched together from a series of bug reports (like how `log_artifact` is patched to return a `_Noop` object, so that people can write `run.log_artifact().wait()`; but the other Artifact-returning methods still return None). Does `log()` really need to update the `summary` for disabled runs? Hyrum's law says someone would have found a way to rely on this.

We could use a more thoughtful design, but that will take a lot more work than just fixing this bug.

### CODE REVIEW

**Comment by timoffex** [comment]
* **#11943** <a href="https://app.graphite.com/github/pr/wandb/wandb/11943?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#11952** <a href="https://app.graphite.com/github/pr/wandb/wandb/11952?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/wandb/wandb/11952?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: wandb/sdk/wandb_init.py [hunk 1]
```diff
@@ -753,94 +753,9 @@ def make_disabled_run(self, config: _ConfigParts) -> Run:
...
-            summary=drun.summary,
-            save=drun.save,
-            use_artifact=drun.use_artifact,
-            log_artifact=drun.log_artifact,
-            define_metric=drun.define_metric,
-            alert=drun.alert,
-            watch=drun.watch,
-            unwatch=drun.unwatch,
+            ),
+            config={**config.sweep_no_artifacts, **config.base_no_artifacts},
```
Comment (dmitryduev) [inline wandb/sdk/wandb_init.py]: I think the order was wrong in the old code, but likely no one cared. Should be base first, then sweeps?

**Review by dmitryduev — APPROVED** [review]
Thank you!

File: wandb/sdk/wandb_init.py (hunk 1 above)
Comment (timoffex) [reply wandb/sdk/wandb_init.py]: Yeah good catch, swapped!

**Comment by timoffex** [comment]
### Merge activity

* **Jun 2, 6:59 PM UTC**: [Graphite](https://app.graphite.com/github/pr/wandb/wandb/11952) rebased this pull request as part of a merge.
* **Jun 2, 7:05 PM UTC**: @timoffex merged this pull request with [Graphite](https://app.graphite.com/github/pr/wandb/wandb/11952).
